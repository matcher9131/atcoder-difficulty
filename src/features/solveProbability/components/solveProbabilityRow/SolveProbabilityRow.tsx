import type { ReactNode } from "react";

export type SolveProbabilityRowProps = {
    readonly contestHeaderCell: ReactNode;
    readonly problemCell: ReactNode;
    readonly difficulty: string;
    readonly solveProbability: string;
};

export const SolveProbabilityRow = ({
    contestHeaderCell,
    problemCell,
    difficulty,
    solveProbability,
}: SolveProbabilityRowProps): ReactNode => {
    return (
        <tr className="contents">
            {contestHeaderCell}
            {problemCell}
            <td className="px-3 py-6 text-right">{difficulty}</td>
            <td className="px-3 py-6 text-right">{solveProbability}</td>
        </tr>
    );
};
