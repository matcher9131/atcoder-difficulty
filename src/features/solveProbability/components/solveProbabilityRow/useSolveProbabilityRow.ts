import type { ProblemSolveProbability } from "../../types/problemSolveProbability";
import type { SolveProbabilityRowProps } from "./SolveProbabilityRow";

export const useSolveProbabilityRow = (problem: ProblemSolveProbability): SolveProbabilityRowProps => {
    // TODO: Implement converting problemId to contestId
    const contestId = "foo";
    const problemId = problem.id;
    const difficulty = `${problem.d?.[1] ?? "NaN"}`;
    const solveProbability =
        problem.solveProbability < 0
            ? "NaN"
            : problem.solveProbability > 0.999
              ? "99.9%"
              : problem.solveProbability < 0.001
                ? "0.1%"
                : `${(problem.solveProbability * 100).toFixed(1)}%`;
    return {
        contestId,
        problemId,
        difficulty,
        solveProbability,
    };
};
