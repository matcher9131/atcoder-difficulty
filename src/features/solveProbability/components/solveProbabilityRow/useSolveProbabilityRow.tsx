import { useAtom } from "jotai";
import { ContestHeaderCellContainer } from "../../../contest/components/contestHeaderCell";
import { ProblemCellContainer } from "../../../problem/components/problemCell";
import { splitProblemId } from "../../../problem/functions/split";
import type { ProblemSolveProbability } from "../../types/problemSolveProbability";
import { toPercent } from "./functions";
import type { SolveProbabilityRowProps } from "./SolveProbabilityRow";
import { selectedProblemAtom } from "../../../distribution/models/selectedProblem";
import { useOpenModalDialog } from "../../../dialog/hooks/useOpenModalDialog";

export const useSolveProbabilityRow = (problem: ProblemSolveProbability): SolveProbabilityRowProps => {
    const [contestId] = splitProblemId(problem.id);
    const contestHeaderCell = <ContestHeaderCellContainer contestId={contestId} />;
    const problemId = problem.id;
    const problemCell = <ProblemCellContainer problemId={problemId} showsParameters={false} showsProblemIndex={true} />;
    const difficulty = problem.d?.[1]?.toString() ?? "NaN";
    const solveProbability = toPercent(problem.solveProbability);
    const [, setSelectedProblem] = useAtom(selectedProblemAtom);
    const { openDialog } = useOpenModalDialog("distribution");
    const handleGraphButtonClick = () => {
        setSelectedProblem(problemId);
        openDialog().catch(console.error);
    };
    return {
        contestHeaderCell,
        problemCell,
        difficulty,
        solveProbability,
        onGraphButtonClick: handleGraphButtonClick,
    };
};
