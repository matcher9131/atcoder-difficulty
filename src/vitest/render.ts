import { render as originalRender, type RenderOptions } from "@testing-library/react";
import userEvent, { type UserEvent } from "@testing-library/user-event";
import type { ReactNode } from "react";

export const render = (
    element: ReactNode,
    options?: Omit<RenderOptions, "queries">,
): ReturnType<typeof originalRender> & { readonly user: UserEvent } => {
    return {
        user: userEvent.setup(),
        ...originalRender(element, options),
    };
};
