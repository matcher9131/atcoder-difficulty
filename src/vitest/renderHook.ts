import { renderHook as originalRenderHook } from "@testing-library/react";
import { Provider } from "jotai";

export const renderHook = <T, Props>(
    render: (initialProps: Props) => T,
): ReturnType<typeof originalRenderHook<T, Props>> => {
    return originalRenderHook(render, { wrapper: Provider });
};
