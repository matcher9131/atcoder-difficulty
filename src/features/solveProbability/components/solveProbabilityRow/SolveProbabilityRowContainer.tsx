import type { ReactNode } from "react";
import { useSolveProbabilityRow } from "./useSolveProbabilityRow";
import type { ProblemSolveProbability } from "../../types/problemSolveProbability";
import { SolveProbabilityRow } from "./SolveProbabilityRow";

type SolveProbabilityRowContainerProps = {
    readonly problem: ProblemSolveProbability;
};

export const SolveProbabilityRowContainer = ({ problem }: SolveProbabilityRowContainerProps): ReactNode => {
    const props = useSolveProbabilityRow(problem);
    return <SolveProbabilityRow {...props} />;
};
