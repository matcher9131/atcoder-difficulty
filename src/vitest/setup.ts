import * as matchers from "@testing-library/jest-dom/matchers";
import type { TestingLibraryMatchers } from "@testing-library/jest-dom/matchers";
import { cleanup } from "@testing-library/react";
import { afterEach, expect } from "vitest";

// Extend Vitest's expect with jest-dom matchers
expect.extend(matchers);

// Merge type declaration of matchers of jest-dom to vitest
declare module "vitest" {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    interface Assertion<T = any> extends jest.Matchers<void, T>, TestingLibraryMatchers<T, void> {}
}

// Clean up the DOM after each test
afterEach(() => {
    cleanup();
});
