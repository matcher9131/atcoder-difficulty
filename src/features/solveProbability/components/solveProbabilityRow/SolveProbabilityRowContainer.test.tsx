import { describe, it, expect, vi, beforeEach } from "vitest";
import { render } from "../../../../vitest/render";
import { screen } from "@testing-library/react";
import { SolveProbabilityRow } from "./SolveProbabilityRow";
import { SolveProbabilityRowContainer } from "./SolveProbabilityRowContainer";
import { useSolveProbabilityRow } from "./useSolveProbabilityRow";

const mockProps = {
    contestHeaderCell: <td data-testid="contest-header">Contest</td>,
    problemCell: <td data-testid="problem-cell">Problem</td>,
    difficulty: "1234",
    solveProbability: "56%",
    onGraphButtonClick: vi.fn(),
};

vi.mock("./useSolveProbabilityRow", () => ({
    useSolveProbabilityRow: vi.fn(),
}));
vi.mock("./SolveProbabilityRow", () => ({
    SolveProbabilityRow: vi.fn(() => <tr data-testid="mock-row" />),
}));

describe("SolveProbabilityRowContainer", () => {
    const sampleProblem = {
        id: "abc",
        n: "Sample Problem",
        d: [1, 100] as [number, number],
        solveProbability: 0.56,
    };

    beforeEach(() => {
        vi.clearAllMocks();
        (useSolveProbabilityRow as ReturnType<typeof vi.fn>).mockReturnValue(mockProps);
    });

    it("calls useSolveProbabilityRow with the correct problem", () => {
        render(<SolveProbabilityRowContainer problem={sampleProblem} />);
        expect(useSolveProbabilityRow).toHaveBeenCalledWith(sampleProblem);
    });

    it("renders SolveProbabilityRow with props from hook", () => {
        render(<SolveProbabilityRowContainer problem={sampleProblem} />);
        expect(SolveProbabilityRow).toHaveBeenCalledWith(expect.objectContaining(mockProps), undefined);
        expect(screen.getByTestId("mock-row")).toBeInTheDocument();
    });

    it("matches snapshot", () => {
        const { container } = render(<SolveProbabilityRowContainer problem={sampleProblem} />);
        expect(container).toMatchSnapshot();
    });
});
