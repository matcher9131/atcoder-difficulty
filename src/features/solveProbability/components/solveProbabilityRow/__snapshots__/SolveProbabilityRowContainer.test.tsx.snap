// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SolveProbabilityRowContainer > matches snapshot 1`] = `
<div>
  <tr
    data-testid="mock-row"
  />
</div>
`;
